### DISCUSSION TITLE
[rustfmt compatibility] Can we expect to auto-format code inside the `view!` macro eventually?

### DISCUSSION THREAD

**Discussion by bluenote10** [leptos-rs/leptos#855] [Q&A] [answered]
Coming from Svelte / React, the feature I miss the most in leptos is to auto-format the code inside the `view!` macro.

Syntax wise, the code inside the view is typically very tedious to format manually, due to its deeply nested nature. Back in the day when I switched from hand-formatting to auto-formatting TSX in React, it was pretty much a game changer: Huge DOM structure refactorings become so much faster, and I could fearlessly move entire TSX fragments out of (or into) a component,  just press `<CTRL>` + `S` and the code snaps into perfect formatting. The thought of having to do all that manually in leptos isn't very encouraging. I need to experiment a bit more with the plain builder syntax, but from what I've seen so far, it may just be a bit too verbose to be as productive as in Svelte/React.

If I understand it correctly rustfmt cannot format macros like the `view!` macro, because it cannot know its syntactical structure, right? On the other hand, rustfmt seems to be able to format some macros like for instance `println!`. What I don't understand: How does it even determine which macros it can and which cannot be formatted?

I'm wondering if there is some hope that this limitation can be lifted or worked around somehow?

**Comment by gbj** [comment, marked as answer]
Check out [`leptosfmt`](https://lib.rs/crates/leptosfmt), which is being built as part of/alongside [`leptos-language-server`](https://github.com/bram209/leptos-language-server). At the moment there's a VS Code extension version of the language server but it's a standard LSP so can be used with other editors too.

**Reply by ChocolateLoverRaj** [reply]
Can we expect to auto-format code inside the `view!` macro eventually without having to install any extensions on Vscode that are specific to Leptos?

**Reply by gbj** [reply]
Generally speaking `rustfmt` treats most proc macros as black boxes and will not format them. You can see this even with something like `cfg-if`, which has a very Rust-like syntax. The choices are:
- use a custom formatter (like `leptosfmt`, which is not VS Code-specific)
- use the builder syntax rather than the macro
- format manually

(Honestly kind of baffled and disheartened by the thumbs down on my comment from last year, which you're replying to.)

If you're aware of another way to support formatting for proc macros in Rust, I'd be excited to hear about it.

**Reply by ChocolateLoverRaj** [reply]
I like the builder syntax. There might be ways to make the builder syntax more convenient like not having to call `.into_view()` for every item in a vec for a fragment.

**Reply by gbj** [reply]
@ChocolateLoverRaj If they're all the same type you can collect an iterator with [`.collect_view()`](https://docs.rs/leptos/latest/leptos/trait.CollectView.html#tymethod.collect_view). If they are not the same type, you can [use a tuple](https://github.com/leptos-rs/leptos/blob/main/examples/counter_without_macros/src/lib.rs#L15-L29) and each one will be converted.

**Reply by bluenote10** [reply]
> Generally speaking rustfmt treats most proc macros as black boxes and will not format them.

Isn't it a bit more subtle? If I remember correctly it does format proc macros if they are written with `(` ... `)` parenthesis, and the content itself is valid Rust syntax or so. I didn't like the idea of using a custom formatting solution very much, which is why I had started working on an alternative macro that can work with plain rustfmt. I didn't had time to follow up on it, but I think it should be possible to have plain rustfmt compatible macro if it sticks to plain Rust syntax.
